highfidelity/hifi
# Fix layered overlay rendering and highlight

https://highfidelity.manuscript.com/f/cases/8296/Layered-drawInFront-drawHUDLayer-Model-Overlays-don-t-render-properly-and-then-crash-on-layer-switch
https://highfidelity.manuscript.com/f/cases/12742/Overlay-in-front-drawn-over-highlighted-object-disappears-because-of-the-highlight-stencil
https://highfidelity.manuscript.com/f/cases/14923/In-VR-lasers-are-hidden-by-highlighted-grabbable-objects-3D-bounding-boxes

Test plan:
- Run [this](https://gist.githubusercontent.com/SamGondelman/c4c8420dee8acf8608ca1db409fa9f6d/raw/21d71bf4e0cf42c38b88216030d29f787e3f33a6/ModelOverlayEntityTest.js).  You should see two rows of models.  The bottom row is entities, and the top row is overlays.  They should look identical, like this:
![hifi-snap-by-samgondelman-on-2018-06-06_10-46-42](https://user-images.githubusercontent.com/7650116/41055852-345f9f48-6977-11e8-9141-4c108bb151ff.jpg)

- Press "u".  The top layer should render "in front" now (for example, in third person, they'll render on top of your avatar).  Besides that, they should still look identical to the bottom row, like this:
![hifi-snap-by-samgondelman-on-2018-06-06_10-46-55](https://user-images.githubusercontent.com/7650116/41055857-39184760-6977-11e8-9a69-c67c13e70dae.jpg)

- Press "i".  The top layer will render in the HUD layer now (for example, if you bring up the stats ("/"), they'll render on top of the 2D overlay).  They'll look darker and not-aliased, but that's ok.  Confirm that they look correct otherwise, like this:
![hifi-snap-by-samgondelman-on-2018-06-06_10-47-04](https://user-images.githubusercontent.com/7650116/41055867-3d5556f6-6977-11e8-8059-51f8e07e8c28.jpg)

- Run [this](https://gist.githubusercontent.com/SamGondelman/5f1dcb5f65d999bc1fdabdb6bb58b147/raw/2532d2844a1f3bdfa9cbcbec881d59df539086d4/LayerOverlayTest.js).  You should see this:
![hifi-snap-by-samgondelman-on-2018-06-06_10-48-05](https://user-images.githubusercontent.com/7650116/41055874-40a2d8ec-6977-11e8-808d-e23324506113.jpg)

- Create a grabbable box.  In VR, when you highlight it with your hand lasers, the lasers should continue rendering properly in front of it.  NOT like this:
https://media.giphy.com/media/3gWxE7QRTlFNaLbYbn/giphy.gif

- Confirm that things still render properly on Android.  Test a few domains.
